presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java

```diff
@@ -398,6 +399,9 @@ public static OrcSelectivePageSource createOrcPageSource(
             }
             catch (IOException ignored) {
             }
+            if (e.getClass().getName().equals(UncheckedExecutionException.class.getName()) && e.getCause() instanceof PrestoException) {
```

Comment:
Use separate catch block instead of checking exception class:

```
        catch (UncheckedExecutionException e) {
            if (e.getCause() instanceof PrestoException) {
                throw (PrestoException) e.getCause();
            }
            throw new PrestoException(HIVE_CANNOT_OPEN_SPLIT, splitError(e, path, start, length), e);
        }
        catch (BlockMissingException e) {
            throw new PrestoException(HIVE_MISSING_DATA, splitError(e, path, start, length), e);
        }
        catch (PrestoException e) {
            throw e;
        }
        catch (Exception e) {
            throw new PrestoException(HIVE_CANNOT_OPEN_SPLIT, splitError(e, path, start, length), e);
        }
        finally {
            try {
                orcDataSource.close();
            }
            catch (IOException ignored) {
            }  
        }
```

Reply:
Catching ```UncheckedExecutionException``` doesn't seem to work here unfortunately :( I think because the thrown exception's class and ```UncheckedExecutionException class``` are loaded by different classloaders. That is why even a simple check like ```e instanceOf UncheckedExecutionException``` doesn't work and we may just have to stick to a class name check.

Just to be sure I tried this on a verifier cluster and it throws the ```HIVE_CANNOT_OPEN_SPLIT``` error like before since it doesn't enter the ```catch (UncheckedExecutionException e) {``` block. (Can share the link and logs in chat if you are interested)

Reply:
And I'm curious about closing the ```orcDataSource```. Doing this in a finally block will close it even in cases when there's no exception, and at a cusrory glance it seems like [callers of this function could still use it](https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java#L391)..
Unrelated to this PR but should we be closing the ```orcDataSource``` in ```OrcSelecticePageSource#close()``` method instead? [here](https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSource.java#L105)

Reply:
@sujay-jain Thanks for explaining. Makes sense. For the benefit of future readers, would you add a comment explaining why you are comparing exception class manually?

Reply:
> And I'm curious about closing the `orcDataSource`. Doing this in a finally block will close it even in cases when there's no exception, and at a cusrory glance it seems like [callers of this function could still use it](https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java#L391)..
> Unrelated to this PR but should we be closing the `orcDataSource` in `OrcSelecticePageSource#close()` method instead? [here](https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSource.java#L105)

@sujay-jain OrcDataSource is closed in AbstractOrcRecordReader.java L380